Make Jmeter able to support nanoseconds for testing and reporting

junyejiang opened the pull request:
In my test experience, when the performance of the system under test is much too fast (for example, when testing Redis), the milliseconds is too large to reflect the performance of the system under test, so the following changes are made:

1. Change the unit of Jmeter Result from milliseconds to nanoseconds.
2. Statistical calculations use BigDecimal/BigInteger to make the statistical results more accurate and no overflow.
3. The test report uses ms/ns as the final output unit according to the Properties configuration.

## Description
<!--- Provide a general summary of your changes in the Title above -->
<!--- Describe your changes in detail here -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Delete as appropriate -->
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [code style][style-guide] of this project.
- [ ] I have updated the documentation accordingly.

[style-guide]: https://wiki.apache.org/jmeter/CodeStyleGuidelines

junyejiang committed:
In my test experience, when the performance of the system under test is much too fast (for example, when testing Redis), the ms is too large to reflect the performance of the system under test, so the following changes are made:
1. Change the unit of Jmeter Result from ms to ns.
2. Statistical calculations use BigDecimal/BigInteger to make the statistical results more accurate and no overflow.
3. The test report uses ms/ns as the final output unit according to the Properties configuration.

vlsi commented on src/jorphan/src/main/java/org/apache/jorphan/math/StatCalculator.java:
```diff
@@ -38,9 +40,9 @@ public abstract class StatCalculator<T extends Number & Comparable<? super T>> {
     // Running values, updated for each sample
-    private double sum = 0;
+    private BigDecimal sum = BigDecimal.ZERO;
 
-    private double sumOfSquares = 0;
+    private BigDecimal sumOfSquares = BigDecimal.ZERO;
 
```
Does this change belong to "support nanoseconds" PR?
If you want to improve precision of `StatCalculator`, I would suggest creating a separate PR and discuss it there.

Have you checked if the change from `double` to `BigDecimal` produces no significant performance regressions?

vlsi commented on src/core/src/test/kotlin/org/apache/jmeter/util/CalculatorTest.kt:
```diff
@@ -66,11 +66,11 @@ class CalculatorTest {
-        calculator.addSample(SampleResult(10, 40))
+        calculator.addSample(SampleResult(10 * 1000000L, 40 * 1000000L))
         assertEquals(40.0, calculator.mean, 0.001, "mean(40)")
-        calculator.addSample(SampleResult(10, 42))
+        calculator.addSample(SampleResult(10 * 1000000L, 42 * 1000000L))
         assertEquals((40.0 + 42.0) / 2, calculator.mean, 0.001, "mean(40, 42)")
-        calculator.addSample(SampleResult(10, 48).apply { sampleCount = 2 })
+        calculator.addSample(SampleResult(10 * 1000000L, 48 * 1000000L).apply { sampleCount = 2 })
```
It does not sound right to modify the existing tests. Please refrain from modifying the existing tests unless it is absolutely necessary.
Otherwise it looks like you break backward compatibilty as we can't require all the plugins to rewrite and recompile their code to add `* 1000000L` all over the place.

vlsi commented on bin/testfiles/HTMLReportTestFile.csv:
```diff
@@ -1,256 +1,256 @@
-1551783686523,267,JR-OK,200,Good request,Thread Group 1-1,text,true,,18,8,1,1,null,0,0,0
-1551783686790,280,JR-OK,200,Good request,Thread Group 1-1,text,true,,18,8,1,1,null,0,0,0
-1551783687071,305,JR-OK,200,Good request,Thread Group 1-1,text,true,,18,8,1,1,null,0,0,0
+1551783627191000000,351000000,JR-OK,200,Good request,Thread Group 1-1,text,true,,18,8,1,1,null,0,0,0
```
Please refrain from modifying the existing tests. Please create a new test for the feature you add.

vlsi commented on bin/reportgenerator.properties:
```diff
@@ -34,6 +34,8 @@
 # Reporting configuration
 #---------------------------------------------------------------------------
 
+jmeter.reportgenerator.msns=ms
```
`jmeter.reportgenerator.msns` is a strange property name. Consider someone would need `minutes` later. I would suggest something like `jmeter.reportgenerator.time.unit=ns|ms|minutes|...`.

However,

vlsi reviewed (changes requested):
I have not checked the way the report works, however, I have a few comments:
1. Have you checked `sampleresult.useNanoTime` JMeter property? If so could you describe why the property is not enough for you?
2. Please refrain from modifying the existing tests. In most cases, if you modify an existing test it means you break backward compatibility which is unwanted.
3. Please create new tests for the feature you add
4. Please refrain from combining multiple unrelated changes into a single commit. I think the change of `double` -> `BigDecimal` should be discussed and measured separately

junyejiang commented:
> I have not checked the way the report works, however, I have a few comments:
> 
> 1. Have you checked `sampleresult.useNanoTime` JMeter property? If so could you describe why the property is not enough for you?
> 2. Please refrain from modifying the existing tests. In most cases, if you modify an existing test it means you break backward compatibility which is unwanted.
> 3. Please create new tests for the feature you add
> 4. Please refrain from combining multiple unrelated changes into a single commit. I think the change of `double` -> `BigDecimal` should be discussed and measured separately

1. Yes I did tried `sampleresult.useNanoTime`,  and finally find out that report was still in ms, this property will not help.
2 & 3. My PR do break backwark compatibility, as the former JTL file was stored in ms format, My PR will change it to ns format. So that the former tests wold not pass until updated.
4. Changes in calculators from Long to BigDecimal is because, firstly every test timestamp is 1 million bigger than before so Long could be easily overflow during accumulation; secondly BigDecimal/BigInteger is more precise than Long in calculation.

junyejiang committed:
Merge branch 'apache:master' into ns

vlsi commented:
>My PR do break backwark compatibility, as the former JTL file was stored in ms format, My PR will change it to ns format. So that the former tests wold not pass until updated.

I am afraid we can't merge such PRs then. We can't break everybody's workflows when releasing a new version.

junyejiang commented:
> > My PR do break backwark compatibility, as the former JTL file was stored in ms format, My PR will change it to ns format. So that the former tests wold not pass until updated.
> 
> I am afraid we can't merge such PRs then. We can't break everybody's workflows when releasing a new version.

Understood. Thanks.

vlsi commented:
It would be great if we could support both formats at the same time, however, keeping the old scripts workable and old files parseable is important

junyejiang commented:
> It would be great if we could support both formats at the same time, however, keeping the old scripts workable and old files parseable is important

I will try to make it. Reopen this PR

junyejiang committed:
Merge branch 'apache:master' into ns

junyejiang committed:
update as CR

junyejiang committed:
Merge branch 'apache:master' into ns

junyejiang commented:
@vlsi 
https://github.com/apache/jmeter/commit/fd12dc39aa688011475c7466cb7c1a2e1f43e38f
Made following changes:
1. Introduced new metadatas in csv, jtl and xml, these metadatas are dedicated to store ns timestamp, old metadatas are untouched.
When Jmeter loads csv, jtl or xml, it tries to read and load ns timestamp first if possible, otherwise it goes the old fasioned way.
2. Added new setter functions in SampleResult.java, these functions are ns timestamp related only, original constructors and factory funcions are untouched.

Above 2 changes are for backward compability.

3. Renamed ns/ms key in reportgenerator.properies.
4. Added test to check my changes.

If my PR is not accepted finally some how, I still suggest to update in calculators from Long to BigInteger/Bigdecimal, which brings better precision and will not case overflow.
